### PR TITLE
feat: Improve namespace validation logic and enhanced inspect output

### PR DESCRIPTION
**Pull request by opjt**
related issue #3851 

@AkihiroSuda 

### fixes
- Added validation to require the --label flag for the update command.
- Added namespace existence checks to both the update and inspect commands.
  - (inspect) If non-existent namespaces are included, a warning log is now emitted. 
    - this logic is based on the behavior`volume-inspect`

### features
- Enhanced namespace inspect output:
  - Added Containers information (count, ids)
  - Added Images information (count, ids)
  - Added Volumes information (count, names)
- The inspect command now provides the same level of information available in namespace list.

### CODE REVIEW

**Comment by Shubhranshu153** [comment]
NIT: Squash the commits. (Probably 2 commits, one for the feature and one for the fix)

File: pkg/cmd/namespace/inspect.go [hunk 1]
```diff
@@ -18,28 +18,134 @@ package namespace
...
+	if err != nil {
+		return info, err
+	}
+
+	info.Count = len(images)
+	ids := make([]string, info.Count)
+
+	info.IDs = ids
+	for idx, img := range images {
+		ids[idx] = strings.Split(img.Target.Digest.String(), ":")[1]
```
Comment (Shubhranshu153) [inline pkg/cmd/namespace/inspect.go]: NIT: Check for validity before accessing the ptr.
Although the digest regex validates the need for an tag, seems safer to do a check.

File: pkg/cmd/namespace/inspect.go [hunk 2]
```diff
@@ -18,28 +18,134 @@ package namespace
...
+		result = append(result, nsInspect)
+	}
+	if err := formatter.FormatSlice(options.Format, options.Stdout, result); err != nil {
+		return err
+	}
+	for _, warn := range warns {
+		log.G(ctx).Warn(warn)
+	}
+
+	if len(warns) != 0 {
```
Comment (Shubhranshu153) [inline pkg/cmd/namespace/inspect.go]: should we throw an error on an warning, my understanding is we should log it and return nil or else cases lets say if there are no volumes it will throw an error.

File: pkg/cmd/namespace/inspect.go [hunk 3]
```diff
@@ -18,28 +18,134 @@ package namespace
...
+			warns = append(warns, err)
+			continue
+		}
+
+		labels, err := namespaceService.Labels(ctx, ns)
 		if err != nil {
 			return err
 		}
+
+		containerInfo, err := containerInfo(ctx, client)
```
Comment (Shubhranshu153) [inline pkg/cmd/namespace/inspect.go:51]: the client and ctx wont have context of the namespaces that we are inspecting as such i dont think this logic is correct.

Can we add some integration tests to make sure we are able to inspect in a namespace that is not the default one which will be picked up. Without tests for edge cases is hard to validate the logic works properly.

File: pkg/cmd/namespace/inspect.go [hunk 4]
```diff
@@ -18,28 +18,134 @@ package namespace
...
+		if err != nil {
+			warns = append(warns, err)
+		}
+
+		imageInfo, err := imageInfo(ctx, client)
```
Comment (Shubhranshu153) [inline pkg/cmd/namespace/inspect.go:56]: same as above for imageInfo.

File: pkg/cmd/namespace/inspect.go (hunk 1 above)
Comment (opjt) [reply pkg/cmd/namespace/inspect.go]: like this ?
``` go
for idx, img := range images {
        digestStr := img.Target.Digest.String()
        if strings.Contains(digestStr, ":") {
	        ids[idx] = img.Target.Digest.Encoded()
        } else {
	        ids[idx] = digestStr
        }
}
```

File: pkg/cmd/namespace/inspect.go (hunk 2 above)
Comment (opjt) [reply pkg/cmd/namespace/inspect.go]: I see we have the same pattern in volume inspect (https://github.com/containerd/nerdctl/blob/main/pkg/cmd/volume/inspect.go#L53). Should we keep the error return for consistency, or do you think both should be changed?

Comment (Shubhranshu153) [reply pkg/cmd/namespace/inspect.go]: There is a slight nuance there, the user is trying to get a particular volume so its correct to return error there.
Here we are trying to inspect the namespace so its ok to not have a volume in that namespace.

File: pkg/cmd/namespace/inspect.go (hunk 1 above)
Comment (Shubhranshu153) [reply pkg/cmd/namespace/inspect.go]: we want the image ID

```
Digest is of the format:
sha256:7173b809ca12ec5dee4506cd86be934c4596dd234ee82c0662eac04a8c2c71dc

splitStr := strings.Split(img.Target.Digest.String(), ":")
if len(splitStr) >= 2 {
    ids[idx] = strings.Split(img.Target.Digest.String(), ":")[1][:12] <short hand is usually 12 characters, this is easy to read>
} else {
     warn <no digest found>
}
```

File: pkg/cmd/namespace/inspect.go (hunk 3 above)
Comment (opjt) [reply pkg/cmd/namespace/inspect.go:51]: I'll work on adding a test case

Comment (opjt) [reply pkg/cmd/namespace/inspect.go:51]: Added test code

File: pkg/cmd/namespace/inspect.go [hunk 5]
```diff
@@ -18,28 +18,135 @@ package namespace
...
+	}
+	for _, warn := range warns {
+		log.G(ctx).Warn(warn)
+	}
+
+	return nil
+}
+func containerInfo(ctx context.Context, client *containerd.Client) (native.ContainerInfo, error) {
+	info := native.ContainerInfo{}
+	containers, err := client.Containers(ctx)
```
Comment (Shubhranshu153) [inline pkg/cmd/namespace/inspect.go:86]: This is still not taking namespace as input (the context and client passed need to be created using the requested namsepace. Same is required for imageInfo

File: cmd/nerdctl/namespace/namespace_inspect_test.go [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+				return helpers.Command("namespace", "inspect", data.Labels().Get("ns1"), data.Labels().Get("ns2"))
+			},
+			Expected: func(data test.Data, helpers test.Helpers) *test.Expected {
+				return &test.Expected{
+					Output: expect.All(
+						expect.Contains(data.Labels().Get("ns1")),
+						expect.JSON([]native.Namespace{}, func(dc []native.Namespace, t tig.T) {
+							assert.Assert(t, len(dc[0].Volumes.Names) == 1, fmt.Sprintf("expected 1 volume (was %d)", len(dc[0].Volumes.Names)))
+							assert.Assert(t, dc[0].Containers.Count == 2, fmt.Sprintf("expected 2 container (was %d)", dc[0].Containers.Count))
+							assert.Assert(t, len(dc[0].Images.IDs) == 1, fmt.Sprintf("expected 1 image (was %d)", dc[0].Images.Count))
```
Comment (Shubhranshu153) [inline cmd/nerdctl/namespace/namespace_inspect_test.go]: These Test from what i understand is not correct, you entered 2 namespace so it should show both the the IDS, same for above.

The logic below is not right so my guess is the tests are needs to be edited as well.

File: pkg/cmd/namespace/inspect.go (hunk 5 above)
Comment (opjt) [reply pkg/cmd/namespace/inspect.go:86]: I seem to be misunderstanding your concern. My recent manual testing of the current implementation shows that the logic correctly filters resources by namespace:

- `nerdctl namespace inspect test1 default` returns accurate counts (e.g., test1: 1 container, default: 2 containers), demonstrating that `containerInfo` and `imageInfo` are correctly scoping results using the `context.Context`.

Could you clarify if you believe the line 
> ctx = namespaces.WithNamespace(ctx, ns)

 is incorrectly used or if `containerd`'s `client.Containers(ctx)` method is known to ignore this context flag?
could you provide an example of a failure scenario or the incorrect output you are observing?

I want to ensure the code is robust against potential future changes, but based on current behavior, the filtering appears to be working as intended by propagating the namespace via the context.

Thanks again for your feedback!

Comment (Shubhranshu153) [reply pkg/cmd/namespace/inspect.go:86]: The client and context for the namespace is created as
```
client, ctx, cancel, err := clientutil.NewClient(cmd.Context(), options.GOptions.Namespace, options.GOptions.Address)
```
in file: https://github.com/containerd/nerdctl/blob/8a5888eaad22007d3f27f43296a6b045172d7c17/cmd/nerdctl/namespace/namespace_inspect.go#L4

so this is taking the global namespace and not the namespace requested in the command.
So for each namespace we are inspecting we need to specify the correct namespace we want to check.

Comment (Shubhranshu153) [reply pkg/cmd/namespace/inspect.go:86]: For the test i would rather trust the id rather than the numbers.
Also run it locally and check you see the different ids correctly.
if it works then your logic is correct.

Comment (opjt) [reply pkg/cmd/namespace/inspect.go:86]: it works
``` bash
pjt@lima-default:/Users/pjt/Projects/PJT/fork-git/nerdctl$ nerdctl namespace inspect test1 default
[
    {
        "Name": "test1",
        "Labels": {
            "tes": "3",
            "test": "1"
        },
        "Containers": {
            "count": 1,
            "ids": [
                "2e7b2558d94acc26ab4f4d8f40d6ffb5460a80a57cc1b0174c92c170b9773d27"
            ]
        },
        "Images": {
            "count": 2,
            "ids": [
                "4b7ce07002c6",
                "ec14c7992a97"
            ]
        },
        "Volumes": {
            "count": 0,
            "names": []
        }
    },
    {
        "Name": "default",
        "Labels": {
            "default": "true"
        },
        "Containers": {
            "count": 2,
            "ids": [
                "973a7436d32452fe754231fd1d68250522ac902108df4b40c8bb78139d04e83a",
                "b07920264d8e9254fb084bef494329e0eb2894893bffdb2e91eef2e6e261df46"
            ]
        },
        "Images": {
            "count": 2,
            "ids": [
                "4b7ce07002c6",
                "f7931603f70e"
            ]
        },
        "Volumes": {
            "count": 1,
            "names": [
                "12"
            ]
        }
    }
]

pjt@lima-default:/Users/pjt/Projects/PJT/fork-git/nerdctl$ nerdctl image ls --namespace test1
REPOSITORY                          TAG         IMAGE ID        CREATED         PLATFORM       SIZE       BLOB SIZE
ghcr.io/stargz-containers/alpine    3.13-org    ec14c7992a97    23 hours ago    linux/amd64    6.218MB    2.814MB
alpine                              latest      4b7ce07002c6    30 hours ago    linux/arm64    9.175MB    4.14MB
pjt@lima-default:/Users/pjt/Projects/PJT/fork-git/nerdctl$ nerdctl image ls --namespace default
REPOSITORY     TAG       IMAGE ID        CREATED       PLATFORM       SIZE       BLOB SIZE
hello-world    latest    f7931603f70e    7 days ago    linux/arm64    12.29kB    4.622kB
alpine         latest    4b7ce07002c6    7 days ago    linux/arm64    9.175MB    4.14MB
```
Is there anything I'm doing wrong?

Verifying by IDs instead of just the count is indeed a better testing. Thank you for the feedback

Comment (Shubhranshu153) [reply pkg/cmd/namespace/inspect.go:86]: its correct, i was missing this line

```
ctx = namespaces.WithNamespace(ctx, ns)
```
the context is updated.

**Comment by opjt** [comment]
@AkihiroSuda When you have time, could you please take a look at this PR?

File: pkg/inspecttypes/native/namespace.go
Comment (AkihiroSuda) [inline pkg/inspecttypes/native/namespace.go:1]: Why include IDs?

File: pkg/cmd/namespace/inspect.go
Comment (AkihiroSuda) [inline pkg/cmd/namespace/inspect.go:1]: This change will make inspect slower, no?

**Comment by AkihiroSuda** [comment]
> features

This seems overloaded to this PR recently?
I suggest splitting this to another PR for further discussion.
By default, the namespace operations should remain lightweight.

**Comment by opjt** [comment]
Thank you for the suggestion.

Regarding `namespace inspect`, I agree that it should contain more details, aligning with the discussion in [3851](https://github.com/containerd/nerdctl/issues/3851) . Is it acceptable to omit the IDs information in the output?

I also agree that this PR is currently overloaded with a mix of fixes and features, which makes it hard to review. I will split the changes into separate, more focused PRs for better reviewability and resubmit them shortly.
